jena-core/src/main/java/org/apache/jena/graph/GraphUtil.java

```diff
@@ -115,55 +117,44 @@ public static void add(Graph graph, Triple[] triples) {
     }
         
     public static void add(Graph graph, List<Triple> triples) {
+        addIteratorWorkerDirect(graph, triples.iterator());
         if ( OldStyle && graph instanceof GraphWithPerform )
-        {
-            GraphWithPerform g = (GraphWithPerform)graph ;
-            for (Triple t : triples)
-                g.performAdd(t) ;
             graph.getEventManager().notifyAddList(graph, triples) ;
-        } else
-        {
-            for (Triple t : triples)
-                graph.add(t) ;
-        }
     }
         
     public static void add(Graph graph, Iterator<Triple> it) {
-        // Materialize to avoid ConcurrentModificationException.
-        List<Triple> s = IteratorCollection.iteratorToList(it) ;
-        if ( OldStyle && graph instanceof GraphWithPerform )
-        {
-            GraphWithPerform g = (GraphWithPerform)graph ;
-            for (Triple t : s)
-                g.performAdd(t) ;
+        if ( OldStyle && graph instanceof GraphWithPerform ) {
+            // Materialize for the notify.
+            List<Triple> s = IteratorCollection.iteratorToList(it) ;
```

Comment:
I'm confused as to why we want to ensure notification this early-- why not wait for the actual work done below? Maybe we could have another line for the comment here on this?